src/zkevm_specs/evm/execution/addmod.py

```diff
@@ -0,0 +1,61 @@
+from ..instruction import Instruction, Transition
+from ..opcode import Opcode
+from zkevm_specs.util import FQ, RLC
+
+
+# Returns 1 when a is lower than b, 0 otherwise
+def lt_u256(instruction: Instruction, a: RLC, b: RLC) -> FQ:
+    # decode RLC to bytes for a and b
+    a_lo, a_hi = instruction.word_to_lo_hi(a, True)
+    b_lo, b_hi = instruction.word_to_lo_hi(b, True)
+
+    a_lt_b_lo, _ = instruction.compare(a_lo, b_lo, 16)
+    a_lt_b_hi, a_eq_b_hi = instruction.compare(a_hi, b_hi, 16)
+
+    a_lt_b = instruction.select(
+        a_lt_b_hi, FQ(1), instruction.select(a_eq_b_hi * a_lt_b_lo, FQ(1), FQ(0))
+    )
+
+    return a_lt_b
+
+
+def addmod(instruction: Instruction):
+
+    opcode = instruction.opcode_lookup(True)
+    instruction.constrain_equal(opcode, Opcode.ADDMOD)
+
+    a = instruction.stack_pop()
+    b = instruction.stack_pop()
+    n = instruction.stack_pop()
+    pushed_r = instruction.stack_push()
+
+    # witness
+    if n.int_value == 0:
+        d = 0
+        r = RLC((a.int_value + b.int_value) % (2**256))
+    else:
+        d = (a.int_value + b.int_value) // n.int_value
+        r = pushed_r
+
+    assert (a.int_value + b.int_value) == (d * n.int_value) + r.int_value
+
+    # check a + b ≡ d * n + r, with carry
+    a_plus_b, left_carry = instruction.add_words([a, b])
+
+    right_carry = instruction.mul_add_words(RLC(d), n, r, a_plus_b)
```

Comment:
I just realized that `mul_add_words` sums 256-bits overflow's 4 limbs together as its overflow, it should only be used to check whether it has overflow or not, instead of being used to compare (there could be many different 256-bits values sharing the same sum of their limbs).

To address worst case, `d` also needs to be up to <code>2 * (2<sup>256</sup>-1)</code> if both `a` and `b` are <code>2<sup>256</sup>-1</code> and `n` is `1`. Then we should make sure 2 values `a+b` and `d*n + r` are equal and both of them have maximum <code>2 * (2<sup>256</sup>-1)</code>.

Reply:
urgh... yes, `2*(2^256-1)` does not fit into `d`  🥴 

I hope that this works... https://github.com/privacy-scaling-explorations/zkevm-specs/pull/189/commits/33847afa3db68511773cad9df72482dfcced7b0d, or maybe I'm missing something again?

Reply:
Hmm I think what we need is (instead of `% (2**256)`):

```python
assert d <= 2*(2**256 - 1)
assert d * n <= 2*(2**256 - 1)
assert d * n + r == a + b # 33 bytes comparison
```

But I understand currently the `Instruction` doesn't have a good API to describe. Just wanted to remind that `d` might not fit into 32 bytes. Perhaps in circuit side we will see this more clearly!